### PR TITLE
ENI: fix calculateExcessIPs excessive calculate of excess ip

### LINKED ISSUE #27970: ENI: calculateExcessIPs not consider minAllocate when usedIP+preAllocate less minAllocate and availableIPs greate than MinAllocate
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I found the operator released more ips than expected, and it needs to allocate ips in the next sync cycle.
It causes unnecessary calls to cloud APIs and the Instability of the IP resource pool.

In this situation need to Release ip:
```
availableIPs:      61,
usedIPs:           29,
preAllocate:       20,
minAllocate:       60,
maxAboveWatermark: 0,
```
release ip count is `availableIPs-usedIPs-preAllocate=61-29-20=12`. But According to the design it only excess 1 IP `availableIPs -minAllocate=61-60=1` 

Next resync cycle after 12 IP is released,  it is needed ip:
```
availableIPs:     49,
usedIPs:          29,
preAllocate:      20,
minAllocate:      60,
maxAllocate:      0,
```
need ip count is `minAllocate-availableIPs=60-49=11`

The test code for this situation is incorrect. But the comment `1 excess` is right. 
https://github.com/cilium/cilium/blob/3a23173e76cf5e379ff58d01cd74c5a2a57765ae/pkg/ipam/node_test.go#L42

### Cilium Version

master--all release version

### Kernel Version

no

### Kubernetes Version

no

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

https://github.com/cilium/cilium/blob/3a23173e76cf5e379ff58d01cd74c5a2a57765ae/pkg/ipam/node.go#L327-L352

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by wu0407**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

If usedIPs+preAllocate not over minAllocate + maxAboveWatermark and availableIPs over minAllocate + maxAboveWatermark, only care  IPs out of minAllocate + maxAboveWatermark calculateExcessIPs.

Fixes: #27970

```release-note
ENI: fix calculateExcessIPs excessive calculate of excess ip
```

### CODE REVIEW

**Comment by gandro** [comment]
I left a comment on the issue if this is really intended: https://github.com/cilium/cilium/issues/27970#issuecomment-1754714778

**Comment by gandro** [comment]
Thanks for your patience in explaining this to me. I agree that this is a bug.

Could you also update the documentation to match the new formula? Thanks!
https://github.com/cilium/cilium/blob/16e4b689daf1f47a429aeae23f4d106bdbb8020d/Documentation/network/concepts/ipam/eni.rst#L295

File: Documentation/network/concepts/ipam/eni.rst [hunk 1]
```diff
@@ -308,15 +308,28 @@ is performed at the interval based scan.
 When determining whether a node has a deficit in IP addresses, the following
 calculation is performed:
 
+``availableIPs``: all IPs on ENI contain used IPs and not used IPs.
```
Comment (gandro) [inline Documentation/network/concepts/ipam/eni.rst]: Why can't we not use `len(spec.ipam.available)` instead?

File: Documentation/network/concepts/ipam/eni.rst [hunk 2]
```diff
@@ -308,15 +308,28 @@ is performed at the interval based scan.
...
+
 .. code-block:: go
 
-     spec.ipam.pre-allocate - (len(spec.ipam.available) - len(status.ipam.used))
+     needIPs = max(spec.ipam.pre-allocate - availableIPs - len(status.ipam.used), spec.ipam.min-allocate - availableIPs)
```
Comment (gandro) [inline Documentation/network/concepts/ipam/eni.rst]: nit:

```suggestion
     neededIPs = max(spec.ipam.pre-allocate - availableIPs - len(status.ipam.used), spec.ipam.min-allocate - availableIPs)
```

**Review by gandro — COMMENTED** [review]
Looks good, one question

File: Documentation/network/concepts/ipam/eni.rst (hunk 1 above)
Comment (wu0407) [reply Documentation/network/concepts/ipam/eni.rst]: The `spec.ipam` field does not contain the `available` attribute. https://github.com/wu0407/cilium/blob/8c77e5b059a80d7a1bcf9a213437ec98cc8ae4d3/pkg/ipam/types/types.go#L121

Comment (gandro) [reply Documentation/network/concepts/ipam/eni.rst]: Ah true, this should be `len(spec.ipam.pool)`, right?

Maybe we can do `availableIP := len(spec.ipam.pool)` to make it a bit more readable

Comment (wu0407) [reply Documentation/network/concepts/ipam/eni.rst]: yeah

File: Documentation/network/concepts/ipam/eni.rst [hunk 3]
```diff
@@ -308,15 +308,29 @@ is performed at the interval based scan.
 When determining whether a node has a deficit in IP addresses, the following
 calculation is performed:
 
+
 .. code-block:: go
 
-     spec.ipam.pre-allocate - (len(spec.ipam.available) - len(status.ipam.used))
+     availableIP := len(spec.ipam.pool)
+     neededIPs = max(spec.ipam.pre-allocate - availableIPs - len(status.ipam.used), spec.ipam.min-allocate - availableIPs)
```
Comment (gandro) [inline Documentation/network/concepts/ipam/eni.rst]: Shouldn't this be the following?

```suggestion
     neededIPs = max(spec.ipam.pre-allocate - (availableIPs - len(status.ipam.used)), spec.ipam.min-allocate - availableIPs)
```

**Review by gandro — COMMENTED** [review]
Thanks! One last thing that I noticed, then this should be good!

**Review by gandro — APPROVED** [review]

**Comment by gandro** [comment]
> I think these failed tests do not relate to this PR. Does it need to be rebased to the main?

Yes, I think recent the `cilium-dbg` rename requires a current branch for CI to pass.

**Comment by wu0407** [comment]
The quay.io is unstable it responds http 500 for push manifest.

**Comment by gandro** [comment]
> The quay.io is unstable it responds http 500 for push manifest.

You will have to re-trigger the build, otherwise other CI workflows will fail due to the missing image. To re-trigger the image build, either do a force-push or close and reopen the PR

**Comment by wu0407** [comment]
I will fix this 
```
level=info msg="Synchronized ENI information" numInstances=1 numSecurityGroups=2 numSubnets=1 numVPCs=1 subsys=eni
level=info msg="Synchronized ENI information" numInstances=1 numSecurityGroups=2 numSubnets=1 numVPCs=1 subsys=eni
level=info msg="Synchronized ENI information" numInstances=1 numSecurityGroups=2 numSubnets=1 numVPCs=1 subsys=eni
level=info msg="Releasing excess IPs from node" available=18 excess=5 excessIps="[10.10.10.7 10.10.111.163 10.10.61.225 10.10.82.173 10.10.53.94]" instanceID=i-testNodeManagerReleaseAddress-1 name=node3 releasing="[10.10.10.7 10.10.111.163 10.10.61.225 10.10.82.173 10.10.53.94]" selectedInterface=75e5dd75-05f7-462c-aba6-33d5d8b00487 selectedPoolID=s-1 subsys=ipam used=10
level=info msg="Synchronized ENI information for the corresponding instance" instance=i-testNodeManagerReleaseAddress-1 numSecurityGroups=2 numSubnets=1 numVPCs=1 subsys=eni
    node_manager_test.go:473: 
        ... obtained int = 13
        ... expected int = 15
        
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃   FAIL  package: github.com/cilium/cilium/pkg/ipam   ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
--- FAIL: Test (13.48s)
--- FAIL: Test/IPAMSuite (13.48s)
--- FAIL: Test/IPAMSuite/TestNodeManagerReleaseAddress (6.04s)
level=info msg="Discovered new CiliumNode custom resource" name=node3 subsys=ipam
level=warning msg="Unable to compute pending pods, will not surge-allocate" error="pod store uninitialized" instanceID= name=node3 subsys=ipam
level=info msg="Resolving IP deficit of node" available=0 availableForAllocation=2048 emptyInterfaceSlots=0 instanceID= maxIPsToAllocate=19 name=node3 neededIPs=15 remainingInterfaces=0 selectedInterface= selectedPoolID=global subsys=ipam used=0
level=warning msg="Unable to compute pending pods, will not surge-allocate" error="pod store uninitialized" instanceID= name=node3 subsys=ipam
level=info msg="Resolving IP deficit of node" available=19 availableForAllocation=2029 emptyInterfaceSlots=0 instanceID= maxIPsToAllocate=8 name=node3 neededIPs=4 remainingInterfaces=0 selectedInterface= selectedPoolID=global subsys=ipam used=19
level=info msg="Releasing excess IPs from node" available=27 excess=8 excessIps="[18 17 16 15 14 13 20 19]" instanceID= name=node3 releasing="[18 17 16 15 14 13 20 19]" selectedInterface= selectedPoolID=global subsys=ipam used=10
    node_manager_test.go:505: 
        ... obtained int = 19
        ... expected int = 18
```

File: Documentation/network/concepts/ipam/eni.rst [hunk 4]
```diff
@@ -308,15 +308,29 @@ is performed at the interval based scan.
 When determining whether a node has a deficit in IP addresses, the following
 calculation is performed:
 
+
```
Comment (learnitall) [inline Documentation/network/concepts/ipam/eni.rst]: Do you mind removing this newline?

File: Documentation/network/concepts/ipam/eni.rst [hunk 5]
```diff
@@ -308,15 +308,29 @@ is performed at the interval based scan.
...
 .. code-block:: go
 
-     spec.ipam.pre-allocate - (len(spec.ipam.available) - len(status.ipam.used))
+     availableIP := len(spec.ipam.pool)
```
Comment (learnitall) [inline Documentation/network/concepts/ipam/eni.rst]:
```suggestion
     availableIPs := len(spec.ipam.pool)
```

File: Documentation/network/concepts/ipam/eni.rst [hunk 6]
```diff
@@ -308,15 +308,29 @@ is performed at the interval based scan.
...
+     if spec.ipam.max-allocate > 0 {
+      neededIPs = min(max(spec.ipam.max-allocate - availableIPs, 0), neededIPs)
+     }
 
 For excess IP calculation:
 
 .. code-block:: go
 
-     (len(spec.ipam.available) - len(status.ipam.used)) - (spec.ipam.pre-allocate + spec.ipam.max-above-watermark)
+     availableIP := len(spec.ipam.pool)
```
Comment (learnitall) [inline Documentation/network/concepts/ipam/eni.rst]:
```suggestion
     availableIPs := len(spec.ipam.pool)
```

**Review by learnitall — CHANGES_REQUESTED** [review]
There seem to be a couple of typos in the docs, but otherwise looks good!

**Review by christarazi — APPROVED** [review]

**Review by learnitall — APPROVED** [review]

**Pull request merged**